compiler/rustc_metadata/src/native_libs.rs

```diff
@@ -433,6 +434,52 @@ impl Collector<'tcx> {
                 }
             }
         };
-        DllImport { name: item.ident.name, ordinal: None, calling_convention, span: item.span }
+
+        DllImport {
+            name: item.ident.name,
+            ordinal: self.extract_ordinal(item),
+            calling_convention,
+            span: item.span,
+        }
+    }
+
+    fn extract_ordinal(&self, foreign_function: &hir::ForeignItemRef<'_>) -> Option<u16> {
+        let mut result: Option<u16> = None;
+
+        for attr in self
+            .tcx
+            .hir()
+            .attrs(foreign_function.id.hir_id())
+            .iter()
+            .filter(|a| a.has_name(sym::link_ordinal))
+        {
+            let items = match attr.meta_item_list() {
+                Some(items) => items,
+                None => {
+                    continue;
+                }
+            };
+
+            if result.is_some() {
+                // FIXME: warn or error on multiple #[link_ordinal] attributes on a single function?
```

Comment:
Personally I think an error is best. It can always be changed to a warning in the future if a use case is discovered but the opposite is not possible.

Reply:
done, but added to rustc_typeck/src/collect.rs.